lib/Reader/abi.cpp

```diff
@@ -125,17 +129,136 @@ void X86_64_Win64::computeSignatureInfo(bool IsManagedCallingConv,
+        switch (Desc.eightByteClassifications[I]) {
+        case SystemVClassificationTypeInteger:
+          Ty = Type::getIntNTy(LLVMContext, Size * 8);
+          RequiredIntRegs++;
+          break;
+
+        case SystemVClassificationTypeSSE:
+          Ty = Size <= 4 ? Type::getFloatTy(LLVMContext)
+                         : Type::getDoubleTy(LLVMContext);
+          RequiredSSERegs++;
+          break;
+
+        case SystemVClassificationTypeIntegerReference:
+          assert(Size == 8);
+
+          // This eightbyte contains a GC reference. Type it as a manged
+          // pointer.
+          Ty = Type::getInt8Ty(LLVMContext)->getPointerTo(1);
+          RequiredIntRegs++;
+          break;
+
+        default:
+          llvm_unreachable("unexpected SysV classification");
+        }
+
+        ABIArgInfo::Expansion Expansion;
+        Expansion.TheType = Ty;
+        Expansion.Offset = Desc.eightByteOffsets[I];
+        Expansions.push_back(Expansion);
+      }
+
+      return ABIArgInfo::getExpand(Expansions);
+    }
+  }
+
+  if (Ty->isIntegerTy()) {
+    assert(Ty->getIntegerBitWidth() <= 64);
+
+    RequiredIntRegs = 1;
+    RequiredSSERegs = 0;
+
+    // RyuJIT requires that all arguments and return values smaller than 32 bits
+    // are zero- or sign-extended.
+    if (IsManagedCallingConv && Ty->getIntegerBitWidth() < 32) {
+      return ABITy.isSigned() ? ABIArgInfo::getSignExtend(Ty)
+                              : ABIArgInfo::getZeroExtend(Ty);
+    }
+
+    return ABIArgInfo::getDirect(Ty);
+  }
+
+  if (Ty->isFloatingPointTy()) {
+    assert(Ty->isFloatTy() || Ty->isDoubleTy());
+    RequiredIntRegs = 0;
+    RequiredSSERegs = 1;
+    return ABIArgInfo::getDirect(Ty);
+  }
+
+  assert(Ty->isPointerTy() || Ty->isVoidTy());
+  RequiredIntRegs = 1;
+  RequiredSSERegs = 0;
+  return ABIArgInfo::getDirect(Ty);
+}
+
+void X86_64_SysV::computeSignatureInfo(
+    LLILCJitContext &Context, bool IsManagedCallingConv, ABIType ResultType,
+    ArrayRef<ABIType> ArgTypes, const DataLayout &DL, ABIArgInfo &ResultInfo,
+    std::vector<ABIArgInfo> &ArgInfos) {
+  uint32_t RequiredIntRegs = 0, RequiredSSERegs = 0;
+  ResultInfo = classify(Context, ResultType, DL, IsManagedCallingConv, true,
+                        RequiredIntRegs, RequiredSSERegs);
+
+  uint32_t FreeIntRegs = ResultInfo.getKind() == ABIArgInfo::Indirect ? 5 : 6;
+  uint32_t FreeSSERegs = 8;
+
+  for (auto &Arg : ArgTypes) {
+    ABIArgInfo Info = classify(Context, Arg, DL, IsManagedCallingConv, false,
+                               RequiredIntRegs, RequiredSSERegs);
+
+    // As per the ABI, if any eightbyte of a struct value does not fit in the
+    // number of available registers, the entire struct value is passed on the
+    // stack.
+    if (Info.getKind() == ABIArgInfo::Expand &&
+        (FreeIntRegs < RequiredIntRegs || FreeSSERegs < RequiredSSERegs)) {
+      RequiredIntRegs = 1;
+      Info = ABIArgInfo::getDirect(Arg.getType());
```

Comment:
It isn't: indeed, this should be setting `RequiredIntRegs` and `RequiredSSERegs` to `0`.